modules/select2_publish/select2_publish.module

```diff
@@ -0,0 +1,64 @@
+<?php
+
+/**
+ * @file
+ * This is the select2_publish module.
+ */
+
+use Drupal\Core\Entity\EntityPublishedInterface;
+
+/**
+ * Implements hook_element_info_alter().
+ */
+function select2_publish_element_info_alter(array &$info) {
+  if (!empty($info['select2'])) {
+    $info['select2']['#pre_render'][] = 'select2_publish_add_status_properties';
+  }
+}
+
+/**
+ * Attach status properties to the render element.
+ */
+function select2_publish_add_status_properties($element) {
+  if ($element['#target_type']) {
+    $entity_manager = \Drupal::entityTypeManager();
+    $entities = $entity_manager->getStorage($element['#target_type'])
+      ->loadMultiple(array_keys($element['#options']));
+
+    foreach ($entities as $id => $entity) {
+      if ($entity instanceof EntityPublishedInterface) {
```

Comment:
Is it possible to check this based on `$element['#target_type']` outside the loop?
In that case all have `data-published` set, or none. Maybe solution for default status could be pulled outside and used in this purpose.